### ISSUE TITLE
Determine best testing tool to use moving forward

### ISSUE THREAD

**Issue by evwilkin** [patternfly/patternfly#2583]
Discuss moving forward with axe vs. IBM a11y testing tools

**Comment by jessiehuff** [comment]
I spent a good bit of time using the Dynamic Assessment Plugin and reviewing some of the issues it found, and honestly, IBM has built an impressive product. It’s clear that they have prioritized accessibility, and I seriously commend them for all of their efforts. I do have a few notes on the tooling though, and would love to continue working with IBM to hopefully improve it!

I took a look at this from the perspective of trying to find an a11y automated testing tool that will find accessibility (WCAG 2 and Section 508) violations. The tool should be easy to read/review the findings, avoid false positives in output, and ideally be open source. Consideration is also being given to how much information and guidance is available to help in correcting the violations.


Some DAP results that I would like to discuss further (however, this isn’t an exhaustive list): 

- The WAI-ARIA widget with **IMG role** must have a nonempty alternate text or a label.
     - Why does this rule still apply if `aria-hidden` is on the element? 
     - Example: arrow icon has aria hidden but getting flagged as a violation- is this a false positive?
- Initial DOM focus should be on its textbox when a **combobox** receives focus 
     - This gets flagged on our documentation pages, but focus does appear to go to the textbook when navigating by keyboard - false positive? 
- Accessible name does not **match or contain the visible label text**
     - We have preview links with a label explaining what clicking on the preview link will do. Technically the visible text is the entire example. Is there a way to filter it for specific instances like this? Any suggestions on this type of use case?
     - Example: `<a class="ws-preview__thumbnail-link" href="/v4/documentation/core/components/alertgroup/toast-alert-group" target="_blank" aria-label="Open fullscreen Toast alert group example">` 
- **Label element** is missing descriptive text that identifies the expect input
     - What determines whether a label is descriptive enough? 
     - Example: Dropdown (split examples) was flagged, but I don’t see what would necessarily fix this based on the issue description when selecting more info.
- **Color contrast**- how is it determined? 
     - Disabled elements do not need to meet criteria, but disabled items are still flagged in the results
          - 1.4.3 Contrast (Minimum): Text or images of text that are part of an inactive user interface component, that are pure decoration, that are not visible to anyone, or that are part of a picture that contains significant other visual content, have no contrast requirement.
          - Example: Wizard back button
     - There were a few flagged for contrast issues that manual testing showed to meet standard - false positive?
          - Example: login page results showed that there was a contrast of 1.00, manual testing showed that contrast was 10.48

Also, are there plans/a timeline about when this tool would be easier for general access rather than logging into IBM and copying an authentication token? I’d love to make this as easy as possible for Red Hatters to start using!

Overall, the tooling is pretty extensive, and I’m interested to see if we can make a few improvements or adjust how we’re using the tool to better our accessibility testing. 
For now, I think that we should continue using axe, follow up with IBM to learn more about our findings and revisit in the future.

**Comment by tombrunet** [comment]
@jessiehuff Some notes...
* The WAI-ARIA widget with IMG role must have a nonempty alternate text or a label.
  * This was a false positive and is now fixed in the 'Preview' archive. We generally ignore issues on things that are `display:none`, but we're still looking at how to handle things that are `aria-hidden`. For example, a color contrast issue on something that is `aria-hidden` is still an issue. Since this SVG pattern of usage is becoming more common, we've put in the fix for this particular rule.
* Initial DOM focus should be on its textbox when a combobox receives focus
  * On first glance, this _looks_ like a false positive, but I need to do some more investigation. `combobox` has undergone a lot of churn in the ARIA spec, and I'm told that those specs still aren't working for people, so ARIA 1.2 is going to revise it again. So, even if it is a false positive, as a heads-up, you'll likely need to make a change to this pattern in the near future.
* Accessible name does not match or contain the visible label text
  * This is a fairly new WCAG 2.1 checkpoint, which is aimed at accessibility issues for users of speech driven ATs. When the text on the screen isn't in the accessible name, those ATs have difficulty finding the correct element. The issues that I saw flagged were all technically valid.
* Label element is missing descriptive text that identifies the expect input
  * The example that I saw was a label wrapping a checkbox, but there was no content for the label. The checkbox itself was labelled via `aria-label`. If you want to keep that label so that that area is clickable without adding content to the screen, move the `aria-label` from the checkbox to the label.
* Color contrast- how is it determined?
  * You are correct that disabled buttons have a different standard, however your wizard button is not tagged as disabled - it is only styled as disabled, which is why it was flagged.
  * I don't see this trigger on the login page. Usually when this flags though, there's some hack with layering and opacity that works in certain browser setups, but isn't actually reflected in the DOM, so it ends up being something like white-on-white in with certain sizes/settings/etc. Again, I'd have to see an instance to give you more information. We do calculate opacity and transparency, so this when this is wrong, it's usually a sign of a weird trick being pulled to color the content, which may not be reliable.

We're in the process of removing the auth token requirement, but that's part of a larger UI redesign. You can put in a token of `00000000-0000-0000-0000-000000000000/00000000-0000-0000-0000-000000000000` and it will work today. The Node tool does not require one.

**Comment by seanforyou23** [comment]
Is there any reason we shouldn't go ahead and close this issue?

**Comment by janwright73** [comment]
Thanks @seanforyou23 for the ping on this issue.  We want to hold onto this just one more week as we are working with the PF leads on documentation for a11y support and tooling.

**Comment by tombrunet** [comment]
FYI - not sure if word made it to your team that these tools are now open source: https://github.com/IBMa/equal-access
